paperqa/agents/models.py

```diff
@@ -0,0 +1,532 @@
+)
+
+from ..version import __version__
+from .prompts import STATIC_PROMPTS
+
+logger = logging.getLogger(__name__)
+
+
+class SupportsPickle(Protocol):
+    def __reduce__(self) -> str | tuple[Any, ...]: ...
+    def __getstate__(self) -> object: ...
+    def __setstate__(self, state: object) -> None: ...
+
+
+def strip_answer(answer: Answer):
+    """Filter out extra items that do not need to be returned to the user. Modifies in place."""
+    new_contexts = [
+        Context(
+            context=c.context,
+            score=c.score,
+            text=Text(
+                text="",
+                **c.text.model_dump(exclude={"text", "embedding", "doc"}),
+                doc=paperqa.Doc(**c.text.doc.model_dump(exclude={"embedding"})),
+            ),
+        )
+        for c in answer.contexts
+    ]
+    answer.contexts = new_contexts
+
+
+class AgentStatus(str, Enum):
+    # FAIL - no answer could be generated
+    FAIL = "fail"
+    # SUCCESS - answer was generated
+    SUCCESS = "success"
+    # TIMEOUT - agent took too long, but an answer was generated
+    TIMEOUT = "timeout"
+    # UNSURE - the agent was unsure, but an answer is present
+    UNSURE = "unsure"
+
+
+class AgentPromptCollection(BaseModel):
+    """Configuration for the agent."""
+
+    model_config = ConfigDict(frozen=True)
+
+    agent_system_prompt: str | None = Field(
+        # Matching https://github.com/langchain-ai/langchain/blob/langchain%3D%3D0.2.3/libs/langchain/langchain/agents/openai_functions_agent/base.py#L213-L215
+        default="You are a helpful AI assistant.",
+        description="Optional system prompt message to precede the below agent_prompt.",
+    )
+
+    # TODO: make this prompt more minimalist, instead improving tool descriptions so
+    # how to use them together can be intuited, and exposing them for configuration
+    agent_prompt: str = (
+        "Answer question: {question}"
+        "\n\nSearch for papers, gather evidence, collect papers cited in evidence then re-gather evidence, and answer."
+        " Gathering evidence will do nothing if you have not done a new search or collected new papers."
+        " If you do not have enough evidence to generate a good answer, you can:"
+        "\n- Search for more papers (preferred)"
+        "\n- Collect papers cited by previous evidence (preferred)"
+        "\n- Gather more evidence using a different phrase"
+        "\nIf you search for more papers or collect new papers cited by previous evidence,"
+        " remember to gather evidence again."
+        " Once you have five or more pieces of evidence from multiple sources, or you have tried a few times, "
+        "call {gen_answer_tool_name} tool. The {gen_answer_tool_name} tool output is visible to the user, "
+        "so you do not need to restate the answer and can simply terminate if the answer looks sufficient. "
+        "The current status of evidence/papers/cost is {status}"
+    )
+    paper_directory: str = Field(
```

Comment:
For the directories, let's always have `str | os.PathLike` to allow for `pathlib` users